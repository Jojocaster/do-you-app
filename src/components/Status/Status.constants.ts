import { StatusType } from './Status.types'

export const StatusLabels: Record<StatusType, string> = {
  [StatusType.OFF]: `We're out.`,
  [StatusType.ON]: `We're on, baby!`,
  [StatusType.LOADING]: `Checking ...`,
  [StatusType.ERROR]: `Checking ...`,
}

export const StatusColours: Record<StatusType, string> = {
  [StatusType.OFF]: `red`,
  [StatusType.ON]: `#27ae60`,
  [StatusType.LOADING]: `orange`,
  [StatusType.ERROR]: `orange`,
}
