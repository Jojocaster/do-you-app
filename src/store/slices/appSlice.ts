import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ColorSchemeName } from 'react-native'
import { getEnv } from '../../constants/Env'

export interface RemoteTheme {
  name: string
  theme: ColorSchemeName | null
  icon: string | null
  playerImage: string | null
}
interface Config {
  assets: string
  defaultTheme: RemoteTheme
  themes: RemoteTheme[]
}
interface AppState {
  // used to manage "what's new" modals based on the version
  whatsNew: string | null
  config: Config | null
  // store last config update
  updatedAt: number | null
}

const initialState: AppState = {
  whatsNew: null,
  config: null,
  updatedAt: null,
}

export const fetchConfig = createAsyncThunk('app/fetchConfig', async () => {
  const url = getEnv('API_URL')
  const response = await fetch(`${url}/config`)
  const data = await response.json()

  return data as Config
})

const appSlice = createSlice({
  initialState,
  name: 'app',
  reducers: {
    updateWhatsNew(state, action: PayloadAction<string>) {
      state.whatsNew = action.payload
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action) => {
      console.log('[fetchConfig]: resolved')
      state.updatedAt = new Date().getTime()
      state.config = action.payload
    })
  },
})

export const { updateWhatsNew } = appSlice.actions
export default appSlice.reducer
