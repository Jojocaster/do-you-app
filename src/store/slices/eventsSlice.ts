import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'

export const fetchEvents = createAsyncThunk('events/fetchEvents', async () => {
  const response = await fetch(
    'https://mahina.app/mahina/app/55559291029?page=1'
  )
  const data = await response.json()

  return data as Events
})

interface EventDate {
  date: string
  day: string
  month: string
}
interface EventData {
  description: string
  id: number
  image: string
  location: string
  price_min: number
  price_max: number | null
  title: string
  ticket_links: {
    link: string
    domain: string
  }[]
  start_date: EventDate
  end_date: EventDate
}
interface Events {
  event_dates: string[]
  events_data: EventData[]
  loading: boolean
  updatedAt: number | null
}

const initialState: Events = {
  event_dates: [],
  events_data: [],
  loading: false,
  updatedAt: null,
}

const eventsSlice = createSlice({
  initialState,
  name: 'event',
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchEvents.fulfilled, (state, action) => {
      state.event_dates = action.payload.event_dates
      state.events_data = action.payload.events_data
      state.loading = false
      state.updatedAt = new Date().getTime()
    })
    builder.addCase(fetchEvents.pending, (state) => {
      state.loading = true
    })
    builder.addCase(fetchEvents.rejected, (state) => {
      state.loading = false
    })
  },
})

export default eventsSlice.reducer
