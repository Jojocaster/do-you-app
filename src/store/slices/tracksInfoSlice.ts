import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TRACKLIST_URL } from '../../constants/Endpoints'

interface TrackInfo {
  album: string
  artist: string
  label: string
  played_datetime: string
  release_date: string
  song_link: string
  timecode: string
  title: string
}

export interface TracksInfoState {
  tracks: TrackInfo[]
  loading: boolean
  lastUpdated: number | null
}

export const fetchTracksInfo = createAsyncThunk(
  'tracksInfo/fetchShowInfo',
  async () => {
    const response = await fetch(TRACKLIST_URL)
    const data = await response.json()
    return data as { tracks: TrackInfo[] }
  }
)

const initialState: TracksInfoState = {
  lastUpdated: null,
  loading: true,
  tracks: [],
}

const tracksInfoSlice = createSlice({
  initialState,
  name: 'tracksInfo',
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      fetchTracksInfo.fulfilled,
      (state, action: PayloadAction<{ tracks: TrackInfo[] }>) => {
        state.tracks = action.payload.tracks
        state.loading = false
        state.lastUpdated = new Date().getTime()
      }
    ),
      builder.addCase(fetchTracksInfo.pending, (state) => {
        state.loading = true
      }),
      builder.addCase(fetchTracksInfo.rejected, (state) => {
        state.loading = false
        state.lastUpdated = new Date().getTime()
      })
  },
})

export default tracksInfoSlice.reducer
