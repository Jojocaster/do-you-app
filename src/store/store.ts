import { combineReducers, configureStore } from '@reduxjs/toolkit'
import playerReducer from './slices/playerSlice'
import scheduleReducer from './slices/scheduleSlice'
import showReducer from './slices/showSlice'
import tracksInfoReducer from './slices/tracksInfoSlice'
import settingsReducer from './slices/settingsSlice'
import storage from 'redux-persist/lib/storage'
import AsyncStore from '@react-native-async-storage/async-storage'
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist'

const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStore,
}

const reducers = combineReducers({
  player: playerReducer,
  schedule: scheduleReducer,
  show: showReducer,
  tracksInfo: tracksInfoReducer,
  settings: settingsReducer,
})

const persistedReducer = persistReducer(persistConfig, reducers)

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
})

let persistor = persistStore(store)
export type RootState = ReturnType<typeof store.getState>
export { persistor }
