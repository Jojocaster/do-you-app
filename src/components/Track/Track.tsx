import { MaterialIcons } from '@expo/vector-icons'
import { format, parseJSON } from 'date-fns'
import { useEffect, useMemo, useState } from 'react'
import { StyleSheet, TouchableOpacity } from 'react-native'
import Colors from '../../constants/Colors'
import useColorScheme from '../../hooks/useColorScheme'
import { TrackInfo } from '../../store/slices/tracksInfoSlice'
import { Button } from '../Button/Button'
import * as Clipboard from 'expo-clipboard'
import { Text, View } from '../Themed'

export const Track: React.FC<{
  active: boolean
  onToggle: (track: string) => void
  track: TrackInfo
}> = ({ active, track, onToggle }) => {
  const [copied, setCopied] = useState(false)
  const theme = useColorScheme()
  const timecode = parseJSON(track.played_datetime)
  const formattedTimecode = format(timecode, 'HH:mm')

  const onCopy = () => {
    try {
      Clipboard.setString(`${track.artist} - ${track.title}`)
    } catch (e) {
      console.log('There was an error copying to clipboard.')
    } finally {
      setCopied(true)
    }
  }

  useEffect(() => {
    setCopied(false)
  }, [active])

  // big performance impact, re-rendering all components added a ~200ms delay
  // useMemo prevents that by only re-rendering the tracks involved
  return useMemo(
    () => (
      <View>
        <View style={styles.container} key={track.played_datetime}>
          <View style={styles.timecode}>
            <Text
              style={{
                fontSize: 14,
              }}
            >
              {formattedTimecode}
            </Text>
          </View>
          <TouchableOpacity
            style={{ flex: 1 }}
            onPress={() => onToggle(track.played_datetime)}
          >
            <View style={{ display: 'flex', flexDirection: 'row' }}>
              <View style={{ marginRight: 5 }}>
                <MaterialIcons
                  name={active ? 'expand-less' : 'expand-more'}
                  size={16}
                  color={Colors[theme].text}
                />
              </View>
              <View style={styles.trackContainer}>
                <Text
                  style={{
                    color: Colors[theme].primary,
                    fontWeight: 'bold',
                    fontSize: 14,
                  }}
                >
                  {track.artist}
                </Text>
                <View>
                  <Text>{track.title}</Text>
                </View>

                <View>
                  {active && (
                    <View style={{ marginTop: 10, marginBottom: 10 }}>
                      <Text style={{ fontSize: 12 }}>
                        Confidence:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.score / 10}/10
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Album:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.album}
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Label:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.label}
                        </Text>
                      </Text>
                      <Text style={{ fontSize: 12 }}>
                        Release:{' '}
                        <Text style={{ fontWeight: 'bold' }}>
                          {track.release_date}
                        </Text>
                      </Text>
                      <View
                        style={styles.separator}
                        lightColor="#eee"
                        darkColor="rgba(255,255,255,0.1)"
                      />
                      <View style={{}}>
                        <Button
                          color={copied ? Colors[theme].primary : undefined}
                          icon={copied ? 'check' : 'content-copy'}
                          onPress={onCopy}
                        >
                          {copied ? `Copied` : 'Copy'}
                        </Button>
                      </View>
                    </View>

                    // <View
                    //   style={[
                    //     styles.codeHighlightContainer,
                    //     styles.homeScreenFilename,
                    //   ]}
                    //   darkColor="rgba(255,255,255,0.05)"
                    //   lightColor="rgba(0,0,0,0.05)"
                    // >
                    //   <Text>Album: {track.album}</Text>
                    //   <Text>Release: {track.release_date}</Text>
                    // </View>
                  )}
                </View>
              </View>
            </View>
          </TouchableOpacity>
        </View>
      </View>
    ),
    [active, theme, copied]
  )
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 10,
  },
  timecode: {
    flexBasis: 60,
  },
  trackContainer: {
    flex: 1,
  },
  codeHighlightContainer: {
    fontSize: 10,
    borderRadius: 3,
    padding: 20,
  },
  homeScreenFilename: {
    marginVertical: 10,
    display: 'flex',
  },
  separator: {
    marginVertical: 10,
    height: 1,
    width: '50%',
  },
})
