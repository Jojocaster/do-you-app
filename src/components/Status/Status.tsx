import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useRef, useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { fetchShowInfo, ShowStatus } from '../../store/slices/showSlice'
import { RootState } from '../../store/store'
import { Text, View } from '../Themed'
import { StatusColours, StatusLabels } from './Status.constants'

export const Status: React.FC = () => {
  const dispatch = useDispatch()
  const { status, lastUpdated } = useSelector((state: RootState) => state.show)
  const timeout = useRef<NodeJS.Timeout>()

  const checkStatus = async () => {
    if (!lastUpdated) {
      dispatch(fetchShowInfo())
    } else {
      const now = new Date().getTime()
      const diff = now - lastUpdated
      const diffInMinutes = diff / (1000 * 60)
      if (diffInMinutes > 1) {
        dispatch(fetchShowInfo())
      }
    }
  }

  useEffect(() => {
    checkStatus()
  }, [])

  // check status every minute if not in loading state
  //TODO: turn into background service, check against schedule to avoid hammering the API
  useEffect(() => {
    if (status !== ShowStatus.LOADING) {
      // reset timeout
      if (timeout.current) {
        clearTimeout(timeout.current)
      }

      timeout.current = setTimeout(() => {
        checkStatus()
      }, 1000 * 60)
    }
  }, [status])

  return (
    <View
      style={{
        paddingTop: 50,
        paddingBottom: 20,
        justifyContent: 'space-between',
        alignItems: 'center',
        backgroundColor: '#121212',
        display: 'flex',
        flexDirection: 'row',
        paddingHorizontal: 20,
      }}
    >
      <Text style={{ fontWeight: 'bold' }}>Status: </Text>
      <View
        style={{
          backgroundColor: 'transparent',
          display: 'flex',
          flexDirection: 'row',

          alignItems: 'center',
        }}
      >
        <Text>{StatusLabels[status]}</Text>
        <MaterialCommunityIcons
          name="adjust"
          size={20}
          style={{ marginLeft: 5 }}
          color={StatusColours[status]}
        />
      </View>
    </View>
  )
}
