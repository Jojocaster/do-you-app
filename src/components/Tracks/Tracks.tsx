import { Octicons } from '@expo/vector-icons'
import { useIsFocused } from '@react-navigation/native'
import React, { useEffect, useRef, useState } from 'react'
import { ActivityIndicator, Animated, FlatList } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import Colors from '../../constants/Colors'
import useColorScheme from '../../hooks/useColorScheme'
import { ShowStatus } from '../../store/slices/showSlice'
import { fetchTracksInfo } from '../../store/slices/tracksInfoSlice'
import { RootState } from '../../store/store'
import { Text, View } from '../Themed'
import { Track } from '../Track/Track'

export const Tracks: React.FC = () => {
  const dispatch = useDispatch()
  const [activeTrack, setActiveTrack] = useState<string>()
  const fadeAnim = useRef(new Animated.Value(1)).current
  const theme = useColorScheme()
  const isFocused = useIsFocused()
  const timeout = useRef<NodeJS.Timeout>()
  const { lastUpdated, loading, tracks } = useSelector(
    (state: RootState) => state.tracksInfo
  )
  const { status } = useSelector((state: RootState) => state.show)

  // setup animation
  useEffect(() => {
    Animated.loop(
      Animated.sequence([
        Animated.timing(fadeAnim, {
          toValue: 0.5,
          duration: 1000,
          useNativeDriver: true,
        }),
        Animated.timing(fadeAnim, {
          toValue: 1,
          duration: 1000,
          useNativeDriver: true,
        }),
      ])
    ).start()
  }, [])

  const onToggle = (track: string) => {
    if (track === activeTrack) {
      setActiveTrack(undefined)
    } else {
      setActiveTrack(track)
    }
  }

  const restartTimer = () => {
    // always clear timeout to prevent leaks
    if (timeout.current) {
      clearTimeout(timeout.current)
    }

    // only poll if show is live & screen is focused
    if (status === ShowStatus.ON && isFocused) {
      timeout.current = setTimeout(() => {
        dispatch(fetchTracksInfo())
      }, 1000 * 60)
    }
  }

  const getTracks = async () => {
    // if tracks were never fetched before, request them straight away
    if (!lastUpdated) {
      dispatch(fetchTracksInfo())
    } else {
      const now = new Date().getTime()
      const diff = now - lastUpdated
      const diffInMinutes = diff / (1000 * 60)
      if (diffInMinutes >= 1) {
        dispatch(fetchTracksInfo())
      } else {
        restartTimer()
      }
    }
  }

  useEffect(() => {
    // fetch tracks when tab is focused
    if (isFocused) {
      getTracks()
    }
  }, [isFocused])

  useEffect(() => {
    // restart timer as soon as new tracks have been loaded
    restartTimer()
  }, [lastUpdated])

  if (loading && !tracks.length) {
    return (
      <View
        testID="loader"
        style={{
          flex: 1,
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <ActivityIndicator color={Colors[theme].primary} />
      </View>
    )
  }

  if (!loading && !tracks.length) {
    return (
      <View
        testID="noTracks"
        style={{
          display: 'flex',
          flexDirection: 'column',
          flex: 1,
          alignItems: 'center',
          justifyContent: 'center',
          width: '100%',
        }}
      >
        <Animated.View style={{ opacity: fadeAnim }}>
          <Octicons name="radio-tower" size={30} color={Colors[theme].tint} />
        </Animated.View>
        <Text style={{ marginTop: 20, fontFamily: 'Lato_900Black' }}>
          No bangers here - yet
        </Text>
      </View>
    )
  }

  return (
    <FlatList
      data={tracks}
      renderItem={(track) => (
        <Track
          active={activeTrack === track.item.played_datetime}
          onToggle={onToggle}
          track={track.item}
        />
      )}
      keyExtractor={(track) => track.played_datetime}
      showsVerticalScrollIndicator={false}
      fadingEdgeLength={100}
      overScrollMode="never"
      style={{ marginTop: 20, width: '100%' }}
    />
  )
}
