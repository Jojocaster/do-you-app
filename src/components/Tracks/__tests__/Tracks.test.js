import * as React from 'react'
import renderer from 'react-test-renderer'
import { Tracks } from '../Tracks'
import { act, fireEvent, render } from '@testing-library/react-native'
import { StatusType } from '../../Status/Status.types'
import { configureStore } from '@reduxjs/toolkit'
import { reducers, store } from '../../../store/store'
import { Provider, useDispatch, useSelector } from 'react-redux'
import { ActivityIndicator, TouchableOpacity } from 'react-native'
import { Track } from '../../Track/Track'
import { useIsFocused } from '@react-navigation/native'
import { fetchTracksInfo } from '../../../store/slices/tracksInfoSlice'

const mockStore = {
  settings: {
    darkTheme: false,
  },
  tracksInfo: { loading: true, tracks: [] },
  show: {
    status: StatusType.OFF,
  },
}

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn((cb) => cb(mockStore)),
}))

jest.mock('../../../hooks/useColorScheme', () => jest.fn(() => 'dark'))
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(() => true),
}))

describe('<Tracks />', () => {
  it('should render', () => {
    const tree = renderer.create(<Tracks />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should display a loader if loading = true', () => {
    const instance = renderer.create(<Tracks />).root
    expect(instance.findByProps({ testID: 'loader' })).toBeDefined()
  })

  describe('getTracks()', () => {
    const dispatchMock = jest.fn()
    useDispatch.mockImplementation(() => dispatchMock)

    describe('when lastUpdated is not defined', () => {
      it('should fetch tracks when focused', () => {
        act(() => {
          const instance = renderer.create(<Tracks />)
        })

        expect(dispatchMock).toHaveBeenCalledWith(expect.any(Function))
      })
    })

    describe('when lastUpdated is defined', () => {
      it('should request tracks if lastUpdated was more than 1 minute ago', () => {
        const currentTimestamp = new Date().getTime()
        const oneMinuteAgo = currentTimestamp - 1000 * 60

        useSelector.mockImplementation((cb) =>
          cb({
            ...mockStore,
            tracksInfo: {
              loading: false,
              tracks: [],
              lastUpdated: oneMinuteAgo,
            },
          })
        )

        act(() => {
          const instance = renderer.create(<Tracks />)
        })
        expect(dispatchMock).toHaveBeenCalled()
      })

      it('should restart the timeout and NOT request tracks if lastUpdated was less than a minute ago', () => {
        dispatchMock.mockClear()
        const currentTimestamp = new Date().getTime()
        const spy = jest.spyOn(window, 'setTimeout')

        useSelector.mockImplementation((cb) =>
          cb({
            ...mockStore,
            tracksInfo: {
              loading: false,
              tracks: [],
              lastUpdated: currentTimestamp,
            },
          })
        )

        act(() => {
          const instance = renderer.create(<Tracks />)
        })
        expect(dispatchMock).not.toHaveBeenCalled()
        expect(spy).toHaveBeenCalled()
      })
    })
  })

  it('should display `no tracks` if loading = false & tracks.length = empty', () => {
    useSelector.mockImplementation((cb) =>
      cb({ ...mockStore, tracksInfo: { loading: false, tracks: [] } })
    )
    const instance = renderer.create(<Tracks />).root
    expect(instance.findAllByProps({ testID: 'loader' }).length).toEqual(0)
    expect(instance.findByProps({ testID: 'noTracks' })).toBeDefined()
  })

  it('should display `no tracks` if loading = false & tracks.length = empty', () => {
    useSelector.mockImplementation((cb) =>
      cb({ ...mockStore, tracksInfo: { loading: false, tracks: [] } })
    )
    const instance = renderer.create(<Tracks />).root
    expect(instance.findAllByProps({ testID: 'loader' }).length).toEqual(0)
    expect(instance.findByProps({ testID: 'noTracks' })).toBeDefined()
  })

  it('should display a list of tracks if defined', () => {
    useSelector.mockImplementation((cb) =>
      cb({
        ...mockStore,
        tracksInfo: {
          loading: false,
          tracks: [
            {
              title: 'foo',
              artist: 'bar',
              played_datetime: '2022-05-09 19:01:43+0100',
            },
          ],
        },
      })
    )
    const instance = renderer.create(<Tracks />).root
    expect(instance.findAllByProps({ testID: 'loader' }).length).toEqual(0)
    expect(instance.findAllByProps({ testID: 'noTracks' }).length).toEqual(0)
    expect(instance.findAllByType(Track).length).toBe(1)
  })
})
