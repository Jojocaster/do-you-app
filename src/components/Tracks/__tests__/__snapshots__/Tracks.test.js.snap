// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tracks /> should render 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#1c1c1c",
      },
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
  testID="loader"
>
  <ActivityIndicator
    animating={true}
    color="#FFDC3B"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
